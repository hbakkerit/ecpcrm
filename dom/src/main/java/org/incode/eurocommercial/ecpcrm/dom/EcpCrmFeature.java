package org.incode.eurocommercial.ecpcrm.dom;

import org.togglz.core.annotation.EnabledByDefault;
import org.togglz.core.annotation.Label;
import org.togglz.core.context.FeatureContext;

public enum EcpCrmFeature implements org.togglz.core.Feature {

    /**
     * Use typesafe queries (rather than JDOQL).
     */
    @Label("Use DataNucleus type-safe queries to query database rather than JDOQL")
    @EnabledByDefault
    useTypeSafeQueries;

    public boolean isActive() {
        return FeatureContext.getFeatureManager().isActive(this);
    }

}
