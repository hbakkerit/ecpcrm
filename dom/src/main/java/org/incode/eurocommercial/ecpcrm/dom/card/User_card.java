/**
 * Copyright 2015-2016 Eurocommercial Properties NV
 * <p>
 * Licensed under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.incode.eurocommercial.ecpcrm.dom.card;

import javax.inject.Inject;

import org.apache.isis.applib.annotation.Action;
import org.apache.isis.applib.annotation.ActionLayout;
import org.apache.isis.applib.annotation.Contributed;
import org.apache.isis.applib.annotation.Mixin;
import org.apache.isis.applib.annotation.PropertyLayout;
import org.apache.isis.applib.annotation.SemanticsOf;

import org.incode.eurocommercial.ecpcrm.dom.user.User;

@Mixin
public class User_card {
    private final User user;

    public User_card(User user) {
        this.user = user;
    }

    @Action(semantics = SemanticsOf.SAFE)
    @ActionLayout(contributed = Contributed.AS_ASSOCIATION)
    @PropertyLayout(named = "Card")
    public Card $$() {
        return cardRepository.findByOwner(user);
    }

    @Inject private CardRepository cardRepository;
}
