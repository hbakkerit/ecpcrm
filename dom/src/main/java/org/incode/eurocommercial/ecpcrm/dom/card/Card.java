/**
 *  Copyright 2015-2016 Eurocommercial Properties NV
 *
 *  Licensed under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.incode.eurocommercial.ecpcrm.dom.card;

import javax.jdo.annotations.Column;
import javax.jdo.annotations.IdentityType;
import javax.jdo.annotations.InheritanceStrategy;
import javax.jdo.annotations.PersistenceCapable;
import javax.jdo.annotations.Queries;
import javax.jdo.annotations.Query;

import org.apache.isis.applib.annotation.DomainObject;
import org.apache.isis.applib.annotation.Editing;
import org.apache.isis.applib.annotation.Property;
import org.apache.isis.applib.util.ObjectContracts;

import org.incode.eurocommercial.ecpcrm.dom.CardStatus;
import org.incode.eurocommercial.ecpcrm.dom.center.Center;

import lombok.Getter;
import lombok.Setter;

@PersistenceCapable(
        identityType = IdentityType.DATASTORE
)
@javax.jdo.annotations.Inheritance(strategy = InheritanceStrategy.NEW_TABLE)
@Queries({
        @Query(
                name = "findByExactNumber", language = "JDOQL",
                value = "SELECT "
                        + "FROM org.incode.eurocommercial.ecpcrm.dom.card.Card "
                        + "WHERE number == :number "),
        @Query(
                name = "findByNumberContains", language = "JDOQL",
                value = "SELECT "
                        + "FROM org.incode.eurocommercial.ecpcrm.dom.card.Card "
                        + "WHERE number.indexOf(:number) >= 0 ")
})
@DomainObject(
        editing = Editing.DISABLED,
        bounded = true
)
public class Card implements Comparable<Card> {

    @Override
    public int compareTo(final Card other) {
        return ObjectContracts.compare(this, other, "email");
    }

    public String title() {
        return getNumber();
    }

    @Column(allowsNull = "false")
    @Property
    @Getter @Setter
    private String number;

    @Column(allowsNull = "false")
    @Property
    @Getter @Setter
    private CardStatus status;

    @Column(allowsNull = "true")
    @Property
    @Getter @Setter
    private String clientId;

    @Column(allowsNull = "false")
    @Property
    @Getter @Setter
    private Center center;

}
