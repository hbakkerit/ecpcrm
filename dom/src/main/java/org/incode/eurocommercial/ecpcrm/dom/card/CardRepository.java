/**
 *  Copyright 2015-2016 Eurocommercial Properties NV
 *
 *  Licensed under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.incode.eurocommercial.ecpcrm.dom.card;

import java.math.BigInteger;
import java.util.List;

import javax.inject.Inject;

import com.google.common.base.Strings;

import org.apache.isis.applib.annotation.DomainService;
import org.apache.isis.applib.annotation.NatureOfService;
import org.apache.isis.applib.annotation.Programmatic;
import org.apache.isis.applib.query.QueryDefault;
import org.apache.isis.applib.services.repository.RepositoryService;

import org.incode.eurocommercial.ecpcrm.dom.CardStatus;
import org.incode.eurocommercial.ecpcrm.dom.center.Center;
import org.incode.eurocommercial.ecpcrm.dom.numerator.Numerator;
import org.incode.eurocommercial.ecpcrm.dom.numerator.NumeratorRepository;

@DomainService(
        nature = NatureOfService.DOMAIN,
        repositoryFor = Card.class
)
public class CardRepository {

    @Programmatic
    public List<Card> listAll() {
        return repositoryService.allInstances(Card.class);
    }

    @Programmatic
    public List<Card> listUnassignedCards() {
        return repositoryService.allMatches(
                new QueryDefault<>(
                        Card.class,
                        "findByOwner",
                        "owner", null));
    }

    @Programmatic
    public List<Card> listEnabledCards() {
        return repositoryService.allMatches(
                new QueryDefault<>(
                        Card.class,
                        "findByStatus",
                        "status", CardStatus.ENABLED));
    }

    @Programmatic
    public Card findByExactNumber(
            final String number
    ) {
        return repositoryService.uniqueMatch(
                new QueryDefault<>(
                        Card.class,
                        "findByExactNumber",
                        "number", number));
    }

    @Programmatic
    public List<Card> findByNumberContains(
            final String number
    ) {
        return repositoryService.allMatches(
                new QueryDefault<>(
                        Card.class,
                        "findByNumberContains",
                        "number", number));
    }

    @Programmatic
    public List<Card> findByCenter(
            final Center center
    ) {
        return repositoryService.allMatches(
                new QueryDefault<>(
                        Card.class,
                        "findByCenter",
                        "center", center));
    }


    @Programmatic
    private Card newCard(
            String number,
            final CardStatus status,
            final Center center
    ) {
        Numerator cardNumerator = numeratorRepository.findOrCreateNumerator("cardNumerator", "%d", BigInteger.ZERO);

        final Card card = repositoryService.instantiate(Card.class);

        if(Strings.isNullOrEmpty(number)) {
            while(!cardNumberIsValid(number)) {
                number = cardNumerator.nextIncrementStr();
            }
        }
        card.setNumber(number);

        card.setStatus(status);
        card.setCenter(center);

        BigInteger num = new BigInteger(number);
        if(num.compareTo(cardNumerator.getLastIncrement()) == 1) {
            cardNumerator.setLastIncrement(num);
        }

        repositoryService.persist(card);
        return card;
    }

    @Programmatic
    public Card findOrCreate(
            final String number,
            final CardStatus status,
            final Center center
    ) {
        Card card = findByExactNumber(number);
        if(card == null) {
            card = newCard(
                    number,
                    status,
                    center
            );
        }
        return card;
    }

    @Programmatic
    public String validateFindOrCreate(final String number, final CardStatus status, final Center center) {
        return Strings.isNullOrEmpty(number) || cardNumberIsValid(number) ? null : "Card number is invalid";
    }

    @Programmatic
    public void delete(final Card card) {
        repositoryService.remove(card);
    }

    @Programmatic
    public boolean cardExists(String cardNumber) {
        return findByExactNumber(cardNumber) != null;
    }

    @Programmatic
    public boolean cardNumberIsValid(String cardNumber) {
        if(Strings.isNullOrEmpty(cardNumber)) {
            return false;
        }
        /* Number has 13 digits */
        if(cardNumber.length() != 13) {
            return false;
        }

        /* Number starts with 20 */
        if(!cardNumber.startsWith("20")) {
            return false;
        }

        /* Number starts with center code */
        // TODO: check center

        /* Checksum */
        int[] digits = cardNumber.chars()
                .map(Character::getNumericValue)
                .toArray();
        int[] multipliers = {1, 3, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3};

        int sum = 0;

        for(int i = 0; i < digits.length - 1; i++) {
            sum += digits[i] * multipliers[i];
        }

        int key = (10 - sum % 10) % 10;

        if(digits[digits.length - 1] != key) {
            return false;
        }

        return true;
    }

    @Inject RepositoryService repositoryService;
    @Inject NumeratorRepository numeratorRepository;

}
