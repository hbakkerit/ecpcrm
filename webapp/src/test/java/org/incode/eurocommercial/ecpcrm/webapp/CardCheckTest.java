/**
 * Copyright 2015-2016 Eurocommercial Properties NV
 * <p>
 * Licensed under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.incode.eurocommercial.ecpcrm.webapp;

import java.util.Random;

import javax.inject.Inject;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;

import org.apache.isis.applib.fixturescripts.FixtureScripts;
import org.apache.isis.applib.services.clock.ClockService;

import org.incode.eurocommercial.ecpcrm.dom.CardStatus;
import org.incode.eurocommercial.ecpcrm.dom.card.Card;
import org.incode.eurocommercial.ecpcrm.dom.game.CardGameRepository;
import org.incode.eurocommercial.ecpcrm.fixture.scenarios.demo.DemoFixture;

import static net.javacrumbs.jsonunit.fluent.JsonFluentAssert.assertThatJson;
import static org.assertj.core.api.Assertions.assertThat;

@Ignore
public class CardCheckTest extends EcpCrmTest {
    @Inject FixtureScripts fixtureScripts;

    @Inject ClockService clockService;

    @Inject CardGameRepository cardGameRepository;


    private DemoFixture fs;
    private Card card;

    @Before
    public void setUp() throws Exception {
        // given
        fs = new DemoFixture();
        fixtureScripts.runFixtureScript(fs, null);

        card = fs.getCards().get(new Random().nextInt(fs.getCards().size()));
        assertThat(card).isNotNull();
    }

    private String endpoint = "card-check";

    private String sendRequest(String cardNumber) throws Exception {
        String request = "{\"card\": \"" + cardNumber +"\", "
                       + "\"origin\": \"borne\"}";
        return super.sendRequest(request, endpoint);
    }

    @Test
    public void when_required_parameter_is_missing_we_expect_302_error() throws Exception {
        // given
        String cardNumber = "";

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
            .node("status").isEqualTo(302);
    }

    @Test
    @Ignore
    /* When the device type is not app and the card number contains 3922 */
    // TODO: Hard to test, need to find a nonexisting valid cardnumber that starts with 3922
    public void when_card_does_not_exist_and_is_outdated_we_expect_319_error() throws Exception {
        // given
        String cardNumber = "";

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
                .node("status").isEqualTo(319);
    }

    @Test
    /* When the card number does not match the required pattern */
    public void when_card_does_not_exist_and_has_invalid_number_we_expect_312_error() throws Exception {
        // given
        String cardNumber = "1";

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
                .node("status").isEqualTo(312);
    }

    @Test
    /* When the card status is "tochange" */
    public void when_card_exists_but_is_outdated_we_expect_319_error() throws Exception {
        // given
        card.setStatus(CardStatus.TOCHANGE);
        String cardNumber = card.getNumber();

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
            .node("status").isEqualTo(319);
    }

    @Test
    /* When the card status is not "enabled" */
    public void when_card_exists_but_is_not_enabled_we_expect_303_error() throws Exception {
        // given
        card.setStatus(CardStatus.DISABLED);
        String cardNumber = card.getNumber();

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
                .node("status").isEqualTo(303);
    }

    @Test
    @Ignore
    // TODO: Not sure how to test this
    public void when_card_exists_but_is_not_the_same_center_as_device_we_expect_317_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(317);
    }

    @Test
    @Ignore
    // TODO: Empty users aren't created for cards. So not sure if this applies.
    public void when_card_does_not_exist_and_a_new_user_cant_be_created_for_it_we_expect_313_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(312);
    }

    @Test
    @Ignore
    // TODO: Have to think of a reason why a new card would be created but can't be linked to an - also new - user.
    // TODO: Besides, see TODO at 312
    public void when_card_does_not_exist_and_a_new_user_is_created_for_it_but_they_cant_be_linked_we_expect_314_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(314);
    }

    @Test
    @Ignore
    // TODO: Can't even find status field in user table, which is what is checked in the source code
    public void when_card_exists_but_card_user_is_invalid_we_expect_304_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(304);
    }

    /* These two give regular responses */

    @Test
    public void when_card_exists_but_cant_play_game_we_expect_sad_response() throws Exception {
        // given
        card.play();
        String cardNumber = card.getNumber();

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
                .node("status").isEqualTo(200);
        assertThatJson(response)
                .node("game").isEqualTo(false);
    }

    @Test
    public void when_card_exists_and_can_play_game_we_expect_happy_response() throws Exception {
        // given
        while(!card.canPlay()) {
            card = fs.getCards().get(new Random().nextInt(fs.getCards().size()));
        }
        String cardNumber = card.getNumber();

        // when
        String response = sendRequest(cardNumber);

        // then
        assertThatJson(response)
                .node("status").isEqualTo(200);
        assertThatJson(response)
                .node("game").isEqualTo(true);
    }

}