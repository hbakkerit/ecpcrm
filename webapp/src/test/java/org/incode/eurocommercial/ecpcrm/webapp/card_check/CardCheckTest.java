/**
 * Copyright 2015-2016 Eurocommercial Properties NV
 * <p>
 * Licensed under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.incode.eurocommercial.ecpcrm.webapp.card_check;

import org.junit.Ignore;
import org.junit.Test;

import org.incode.eurocommercial.ecpcrm.webapp.ecp_crm_test.EcpCrmTest;

import static net.javacrumbs.jsonunit.fluent.JsonFluentAssert.assertThatJson;

@Ignore
public class CardCheckTest extends EcpCrmTest {

    private String endpoint = "card-check";

    private String sendRequest(String cardNumber) throws Exception {
        String request = "{\"card\": \"" + cardNumber +"\", "
                       + "\"origin\": \"borne\"}";
        return super.sendRequest(request, endpoint);
    }

    @Test
    public void when_required_parameter_is_missing_we_expect_302_error() throws Exception {
        String cardNumber = "";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(302);
    }

    @Test
    @Ignore
    /* When the device type is not app and the card number contains 3922 */
    // TODO: Hard to test, need to find a nonexisting valid cardnumber that starts with 3922
    public void when_card_does_not_exist_and_is_outdated_we_expect_319_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(319);
    }

    @Test
    /* When the card number does not match the required pattern */
    public void when_card_does_not_exist_and_has_invalid_number_we_expect_312_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(312);
    }

    @Test
    /* When the card status is "tochange" */
    public void when_card_exists_but_is_outdated_we_expect_319_error() throws Exception {
        String cardNumber = "3922071977845";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(319);
    }

    @Test
    /* When the card status is not "enabled" */
    public void when_card_exists_but_is_not_enabled_we_expect_303_error() throws Exception {
        String cardNumber = "2037000090418";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(303);
    }

    @Test
    @Ignore
    // TODO: Not sure how to test this
    public void when_card_exists_but_is_not_the_same_center_as_device_we_expect_317_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(317);
    }

    @Test
    @Ignore
    // TODO: Need to find an unused valid card number, which is not too easy by itself, but also for some reason a user can't be created for it
    public void when_card_does_not_exist_and_a_new_user_cant_be_created_for_it_we_expect_313_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(312);
    }

    @Test
    @Ignore
    // TODO: Have to think of a reason why a new card would be created but can't be linked to an - also new - user
    public void when_card_does_not_exist_and_a_new_user_is_created_for_it_but_they_cant_be_linked_we_expect_314_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(314);
    }

    @Test
    @Ignore
    // TODO: Can't even find status field in user table, which is what is checked in the source code
    public void when_card_exists_but_card_user_is_invalid_we_expect_304_error() throws Exception {
        String cardNumber = "1";
        assertThatJson(sendRequest(cardNumber))
            .node("status").isEqualTo(304);
    }

    /* These two give regular responses */

    // TODO: These tests depend on whether a game was already played this day for the user, so we have to figure out a test case
    @Test
    @Ignore
    public void when_card_exists_but_cant_play_game_we_expect_sad_response() throws Exception {
    }

    @Test
    @Ignore
    public void when_card_exists_and_can_play_game_we_expect_happy_response() throws Exception {
    }


}