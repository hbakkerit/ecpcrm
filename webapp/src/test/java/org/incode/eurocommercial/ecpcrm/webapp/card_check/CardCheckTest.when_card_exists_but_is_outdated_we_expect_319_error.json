{"card":"3922071977845","origin":"borne"}