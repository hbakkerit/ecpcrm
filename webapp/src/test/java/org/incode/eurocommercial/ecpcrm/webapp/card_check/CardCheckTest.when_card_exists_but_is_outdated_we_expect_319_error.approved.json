{
    "status": 319,
    "message": "Cette carte est pÃ©rimÃ©e, vous devez la changer"
}